src/hera/workflows/steps.py

```diff
@@ -38,20 +40,37 @@ class Step(
     with_param: Optional[str]
     with_sequence: Optional[_ModelSequence]
 
-    def _build_as_workflow_step(self) -> _ModelWorkflowStep:
-        # Convert the unified list of Artifacts and Parameters in self.arguments to a
-        # model-class `Arguments`` to pass to WorkflowStep
-        artifacts = list(filter(lambda arg: isinstance(arg, _ModelArtifact), self.arguments or []))
-        parameters = list(filter(lambda arg: isinstance(arg, _ModelParameter), self.arguments or []))
+    def _build_arguments(self) -> Optional[_ModelArguments]:
+        if self.arguments is None:
+            return None
+
+        artifacts = None
+        for arg in self.arguments:
+            if isinstance(arg, _ModelArtifact):
+                artifacts = [arg] if artifacts is None else [*cast(List[_ModelArtifact], artifacts), arg]
+            elif isinstance(arg, Artifact):
+                artifacts = [arg._build_artifact()] if artifacts is None else [*artifacts, arg._build_artifact()]
```

Comment:
IMO we can make this a bit cleaner by intialising `artifacts` to a list:
```suggestion
        artifacts = []  # may need a type hint as List[_ModelArtifact]
        for arg in self.arguments:
            if isinstance(arg, _ModelArtifact):
                artifacts.append(arg)
            elif isinstance(arg, Artifact):
                artifacts.append(arg._build_artifact())
```

And in the assignment of `model_arguments` use `artifacts=None if not artifacts else artifacts` to make use of zero-length-list's Falsey value

[Also applicable to `parameters` below]